src/Pos/Generator/Block/Payload.hs

```diff
@@ -0,0 +1,199 @@
+          if fromIntegral (length res) /= n
+              then error "selectDistinct is broken"
+              else pure res
+  where
+    n :: Int
+    n = min (b + 1 - a) n0
+    selectDistinct' :: [Int] -> Int -> m [Int]
+    selectDistinct' cur leftN = do
+        leftItems <- take leftN <$> getRandomRs (a, b)
+        let nubbed = ordNub $ cur ++ leftItems
+        if length nubbed < n
+            then selectDistinct' nubbed (n - length nubbed)
+            else pure nubbed
+
+-- | Separates coin into provided number of coins. All resulting coins
+-- are nonzero.
+splitCoins :: (MonadRandom m, MonadIO m) => Int -> Coin -> m [Coin]
+splitCoins n (fromIntegral . coinToInteger -> c)
+    | c < n = error $ "splitCoins: can't split " <> pretty c <>
+                      " on " <> show n <> " parts"
+    | otherwise = do
+          splitPoints <- sort <$> selectDistinct n (0, c - 1)
+          -- calculate length of intervals
+          let amounts = map (\(a,b) -> b - a + 1) $ (0 : splitPoints) `zip` splitPoints
+          -- here we can use unsafeIntegerToCoin, because amount of
+          -- subcoin is less than 'c' by design.
+          pure $ map (unsafeIntegerToCoin . fromIntegral) amounts
+
+-- | State datatype for transaction payload generation
+data GenTxData = GenTxData
+    { _gtdUtxo     :: Utxo
+      -- ^ Utxo as it is.
+    , _gtdUtxoKeys :: V.Vector TxIn
+      -- ^ Keys of 'gtdUtxo' to support random selection by index.
+    }
+
+makeLenses ''GenTxData
+
+instance (Monad m) => MonadUtxoRead (StateT GenTxData m) where
+    utxoGet txIn = uses gtdUtxo $ M.lookup txIn
+
+instance (Monad m) => MonadUtxo (StateT GenTxData m) where
+    utxoPut txIn txOutAux = gtdUtxo . at txIn ?= txOutAux
+    utxoDel txIn = gtdUtxo . at txIn .= Nothing
+
+-- TODO: move to txp, think how to unite it with 'Pos.Arbitrary.Txp'.
+-- | Generate valid 'TxPayload' using current global state.
+genTxPayload ::
+       forall g m. (RandomGen g, MonadBlockGenBase m)
+    => BlockGenRandMode g m ()
+genTxPayload = do
+    utxo <- lift DB.getAllPotentiallyHugeUtxo
+    let gtd = GenTxData utxo (V.fromList $ M.keys utxo)
+    flip evalStateT gtd $ do
+        (a,d) <- lift $ view tgpTxCountRange
+        txsN <- fromIntegral <$> getRandomR (a, a + d)
+        void $ replicateM txsN genTransaction
+  where
+    genTransaction :: StateT GenTxData (BlockGenRandMode g m) ()
+    genTransaction = do
+        utxoSize <- uses gtdUtxoKeys V.length
+        when (utxoSize == 0) $
+            -- Empty utxo means misconfigured testing settings most
+            -- probably, so that's a critical error
+            error "Utxo is empty already which is weird"
+
+        secrets <- view asSecretKeys <$> view blockGenParams
+        -- Unsafe hashmap resolving is used here because we suppose
+        -- utxo contains only related to these secret keys only.
+        let resolveSecret stId =
+                fromMaybe (error $ "can't find stakeholderId " <>
```

Comment:
I consider tests as an executable that runs only once. Semantics for `error` is "something is misconfigured and we should fail". If tests are misconfigured and exit because of error -- it's good. There's no reason to catch this exception and process it (hspec does it anyway).